fpsim/methods.py

```diff
@@ -0,0 +1,253 @@
+"""
+Contraceptive methods
+
+Idea:
+    A method selector should be a generic class (an intervention?). We want the existing matrix-based
+    method and the new duration-based method to both be instances of this.
+
+"""
+
+# %% Imports
+import numpy as np
+import sciris as sc
+import pandas as pd
+import starsim as ss  # TODO add to dependencies
+from . import utils as fpu
+from . import defaults as fpd
+from . import locations as fplocs
+
+__all__ = ['Method', 'Methods', 'ContraceptiveChoice', 'RandomChoice', 'SimpleChoice', 'EmpoweredChoice']
+
+
+# %% Base definition of contraceptive methods -- can be overwritten by locations
+class Method:
+    def __init__(self, name=None, label=None, idx=None, efficacy=None, modern=None, dur_use=None, csv_name=None):
+        self.name = name
+        self.label = label or name
+        self.csv_name = csv_name or label or name
+        self.idx = idx
+        self.efficacy = efficacy
+        self.modern = modern
+        self.dur_use = dur_use
+
+
+# Helper function for setting lognormals
+def ln(a, b): return dict(dist='lognormal', par1=a, par2=b)
+
+
+method_list = [
+    Method(name='none',     efficacy=0,     modern=False, dur_use=ln(2, 3), label='None'),
+    Method(name='pill',     efficacy=0.945, modern=True,  dur_use=ln(2, 3), label='Pill'),
+    Method(name='iud',      efficacy=0.986, modern=True, dur_use=ln(5, 3), label='IUDs', csv_name='IUD'),
+    Method(name='inj',      efficacy=0.983, modern=True, dur_use=ln(2, 3), label='Injectables', csv_name='Injectable'),
+    Method(name='cond',     efficacy=0.946, modern=True,  dur_use=ln(1, 3), label='Condoms', csv_name='Condom'),
+    Method(name='btl',      efficacy=0.995, modern=True, dur_use=ln(50, 3), label='BTL', csv_name='F.sterilization'),
+    Method(name='wdraw',    efficacy=0.866, modern=False, dur_use=ln(1, 3), label='Withdrawal', csv_name='Withdrawal'), #     # 1/2 periodic abstinence, 1/2 other traditional approx.  Using rate from periodic abstinence
```

Comment:
The 1/2 periodic absintence 1/2 other traditional is, I think, a  comment for line 47